derive_miniconf/src/lib.rs

```diff
@@ -49,162 +37,157 @@ impl TypeDefinition {
 pub fn derive(input: TokenStream) -> TokenStream {
     let input = parse_macro_input!(input as DeriveInput);
 
-    let typedef = TypeDefinition::new(input.generics, input.ident);
-
     match input.data {
-        syn::Data::Struct(struct_data) => derive_struct(typedef, struct_data),
+        syn::Data::Struct(_) => derive_struct(input),
         _ => unimplemented!(),
     }
 }
 
-/// Derive the Miniconf trait for structs.
-///
-/// # Args
-/// * `typedef` - The type definition.
-/// * `data` - The data associated with the struct definition.
-/// * `atomic` - specified true if the data must be updated atomically. If false, data must be
-///   set at a terminal node.
-///
-/// # Returns
-/// A token stream of the generated code.
-fn derive_struct(mut typedef: TypeDefinition, data: syn::DataStruct) -> TokenStream {
-    let raw_fields = match data.fields {
-        syn::Fields::Named(syn::FieldsNamed { ref named, .. }) => named,
-        _ => unimplemented!("Only named fields are supported in structs."),
-    };
-
-    let fields: Vec<StructField> = raw_fields
-        .iter()
-        .map(|x| StructField::new(x.clone()))
-        .collect();
-
-    for field in fields.iter() {
-        field.bound_generics(&mut typedef);
-    }
-
-    let set_path_arms = fields.iter().map(|f| {
-        let match_name = &f.field.ident;
-        if f.deferred {
-            quote! {
-                stringify!(#match_name) => {
-                    self.#match_name.set_path(path_parts, value)
+fn get_path_arm(f: &StructField) -> proc_macro2::TokenStream {
+    let match_name = &f.field.ident;
+    if f.deferred {
+        quote! {
+            stringify!(#match_name) => {
+                self.#match_name.get_path(path_parts, value)
+            }
+        }
+    } else {
+        quote! {
+            stringify!(#match_name) => {
+                if peek {
```

Comment:
While this is efficient, it's not great from a code maintainability perspective. The `peek` variable is defined pretty far away from the rest of this code, so it may not be clear where it's supposed to be coming from.

IMO we should likely just rely on Rust being smart enough to optimize the redundant `peek()` calls. I'm sure it can do it without issue if the field isn't mutating.

Reply:
The same argument (readability) holds for `value`, `path_parts`, `self` and the rest of the control context.

Reply:
Fair point